### PR TITLE
SONARPY-944 Use precomputed Typeshed symbols for third-party libraries in the Python analyzer

### PR DESCRIPTION
**Pull request by andrea-guarino-sonarsource**

### CODE REVIEW

File: python-frontend/src/main/java/org/sonar/python/types/TypeShed.java [hunk 1]
```diff
@@ -73,6 +72,7 @@
   private static final String THIRD_PARTY_3 = "typeshed/third_party/3/";
   private static final String CUSTOM_THIRD_PARTY = "custom/";
   private static final String PROTOBUF = "protobuf/";
+  private static final String PROTOBUF_THIRD_PARTY = "protobuf/stubs/";
```
Comment (guillaume-dequenne-sonarsource) [inline python-frontend/src/main/java/org/sonar/python/types/TypeShed.java:75]: Currently we have stdlib under `protobuf` and third parties under `protobuf/stubs`. All of this is under `types` which currently holds `Typeshed` as well.
What about going for a `protobuf_stdlib`, `protobuf_thid_parties` and `protobuf_custom` all at the same level, for better clarity?

File: python-checks/src/test/resources/checks/hotspots/cors/cors.py [hunk 2]
```diff
@@ -138,8 +138,7 @@ def flask_add_headers_no_args():
 @app.route("/")
 def flask_make_response_headers_set():
     resp = make_response("hello")
-    resp.headers['Access-Control-Allow-Origin'] = '*' # Noncompliant
-#   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+    resp.headers['Access-Control-Allow-Origin'] = '*' # FN
```
Comment (guillaume-dequenne-sonarsource) [inline python-checks/src/test/resources/checks/hotspots/cors/cors.py:141]: Do you know why this becomes a FN? Is this the one related to the other ticket you created?

File: python-frontend/typeshed_serializer/serializer/typeshed_serializer.py [hunk 3]
```diff
@@ -136,7 +136,7 @@ def save_merged_symbols(is_debug=False, is_third_parties=False):
 
 def main():
     save_merged_symbols()
-    save_merged_symbols(is_third_parties=True)
```
Comment (guillaume-dequenne-sonarsource) [inline python-frontend/typeshed_serializer/serializer/typeshed_serializer.py]: I guess this change is not needed.

**Review by guillaume-dequenne-sonarsource — CHANGES_REQUESTED** [review]
Looks good! I left some minor comments. I think it would be nice merging this in a separate branch so that I can incorporate my changes from SONARPY-945 there as well, so that we can validate every regression in sonar-security is under control before merging everything.

File: python-frontend/typeshed_serializer/serializer/symbols.py [hunk 4]
```diff
@@ -515,9 +519,10 @@ def extract_return_type(func_def: mpn.FuncDef):
     return TypeDescriptor(func_type.ret_type)
 
 
-def save_module(ms: Union[ModuleSymbol, MergedModuleSymbol], is_debug=False, debug_dir="output"):
+def save_module(ms: Union[ModuleSymbol, MergedModuleSymbol], is_debug=False, debug_dir="output", is_stdlib=True):
```
[the remark this answers is not in the document]
Comment (andrea-guarino-sonarsource) [reply python-frontend/typeshed_serializer/serializer/symbols.py:522]: Sure, makes sense to me. I will let you do it in your PR

File: python-frontend/typeshed_serializer/serializer/symbols.py [hunk 5]
```diff
@@ -91,9 +91,13 @@ def __init__(self, _type: mpt.Type):
         elif isinstance(_type, mpt.TupleType):
             self.kind = TypeKind.TUPLE
             items = [TypeDescriptor(t) for t in _type.items]
-            item_names = [i.pretty_printed_name for i in items]
-            self.args.extend(items)
-            self.pretty_printed_name = f"Tuple[{','.join(item_names)}]"
+            if any(item.is_unknown for item in items):
+                self.kind = None
+                self.is_unknown = True
```
[the remark this answers is not in the document]
Comment (andrea-guarino-sonarsource) [reply python-frontend/typeshed_serializer/serializer/symbols.py:96]: Indeed, I hesitated a bit if creating a Tuple with `Any`. I took the simplest approach for the time being, if that's fine also for you, I would prefer to handle this in a separate PR. I can create a ticket for it.

File: python-frontend/typeshed_serializer/serializer/typeshed_serializer.py [hunk 6]
```diff
@@ -60,13 +62,34 @@ def load_single_module(module_fqn: str, category="stdlib"):
...
+    return source_paths, build_result
+
+
+def get_sources(relative_path: str, generate_python2: bool):
+    source_list = []
+    source_paths = set()
     path = os.path.join(CURRENT_PATH, relative_path)
     for root, dirs, files in os.walk(path):
         package_name = root.replace(path, "").replace("\\", ".").replace("/", ".").lstrip(".")
-        if not generate_python2_stdlib and "python2" in package_name:
```
[the remark this answers is not in the document]
Comment (andrea-guarino-sonarsource) [reply python-frontend/typeshed_serializer/serializer/typeshed_serializer.py:69]: Yes Indeed I changed it to generalize. However, also third_parties has `@python2` (for example [six](https://github.com/python/typeshed/tree/2c97c3cab9cdd534744f5b917b39d4f0528c8cdd/stubs/six/%40python2/six))

File: python-frontend/typeshed_serializer/serializer/typeshed_serializer.py [hunk 7]
```diff
@@ -92,7 +115,7 @@ def serialize_typeshed_stdlib(output_dir_name="output", python_version=(3, 8), i
     """
     output_dir_name = output_dir_name if python_version >= (3, 0) else f"{output_dir_name}@python2"
     opt = get_options(python_version)
-    build_result = walk_typeshed_stdlib(opt)
+    _, build_result = walk_typeshed_stdlib(opt)
```
[the remark this answers is not in the document]
Comment (andrea-guarino-sonarsource) [reply python-frontend/typeshed_serializer/serializer/typeshed_serializer.py:118]: Sure, I can do that

File: python-frontend/src/main/java/org/sonar/python/types/TypeShed.java [hunk 8]
```diff
@@ -311,8 +308,9 @@ private static boolean isAmbiguousSymbolOfClasses(Symbol symbol) {
     return false;
   }
 
-  private static Map<String, Symbol> getSymbolsFromProtobufModule(String moduleName) {
-    InputStream resource = TypeShed.class.getResourceAsStream(PROTOBUF + moduleName + ".protobuf");
+  private static Map<String, Symbol> getSymbolsFromProtobufModule(String moduleName, boolean isThirdParty) {
```
[the remark this answers is not in the document]
Comment (andrea-guarino-sonarsource) [reply python-frontend/src/main/java/org/sonar/python/types/TypeShed.java:311]: As discussed, I will let you do that in your PR :)

File: python-frontend/typeshed_serializer/serializer/symbols.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (guillaume-dequenne-sonarsource) [reply python-frontend/typeshed_serializer/serializer/symbols.py:96]: Sure!

File: python-checks/src/test/resources/checks/hotspots/cors/cors.py (hunk 2 above)
Comment (andrea-guarino-sonarsource) [reply python-checks/src/test/resources/checks/hotspots/cors/cors.py:141]: This is because attributes were missing from typeshed classes. I think this change is not needed after I rebase on master :)

File: python-frontend/src/main/java/org/sonar/python/types/TypeShed.java (hunk 1 above)
Comment (andrea-guarino-sonarsource) [reply python-frontend/src/main/java/org/sonar/python/types/TypeShed.java:75]: works for me! if that's fine I will let you also do that in your PR or I can do it in another separate PR

File: python-frontend/typeshed_serializer/serializer/typeshed_serializer.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (andrea-guarino-sonarsource) [reply python-frontend/typeshed_serializer/serializer/typeshed_serializer.py]: Done in 75f135360536e65c81d5ab4b8524b06cd7791ae9

File: python-checks/src/test/resources/checks/hotspots/cors/cors.py (hunk 2 above)
Comment (andrea-guarino-sonarsource) [reply python-checks/src/test/resources/checks/hotspots/cors/cors.py]: I confirm, after rebase the issue is raised again :)

File: python-frontend/typeshed_serializer/serializer/typeshed_serializer.py (hunk 3 above)
Comment (andrea-guarino-sonarsource) [reply python-frontend/typeshed_serializer/serializer/typeshed_serializer.py]: Fixed in 90ef511687a014c251b8a76215d7303933287b15

**Review by guillaume-dequenne-sonarsource — APPROVED** [review]
LGTM
